Consolidate indexer run time tracking in Indexer class (mostly)

hannes-ucsc opened the pull request:
So far we've been neglecting to accurately track the Lambda invocation's remaining run time in the indexer. No one noticed until we started retrying to load metadata blobs in #1137 as a fix for #1115. The retries could potentially take so much time that there wouldn't be anything left for the actual indexing operation, like interacting with ES. We should make reasonably sure that there is sufficient time to complete an indexing operation before we commence it, because if we are interrupted, we'll leave the index inconsistent. This PR fixes that on a best effort basis. I've opened #1164 to retry the entire invocation if that best effort fails.

xbrianh commented on daemons/dss-index/app.py:
```diff
@@ -42,14 +42,12 @@ def dispatch_gs_indexer_event(event, context):
     # We can't use ecxecutor as context manager because we don't want the shutdown to block
     try:
-        backend = CompositeIndexBackend(executor, DEFAULT_BACKENDS, timeout=timeout, context=context)
+        backend = CompositeIndexBackend(executor, DEFAULT_BACKENDS)
```
Is `executor` only used in `backend._delegate`?  I'm curious why it's passed in here.

hannes-ucsc replied:
`_delegate` is a protected method that should not be called from outside the class. The backend methods that call it are public but their signature is restricted by the abstract class (BTW: this is an instance of the Composite design pattern). The lifetimes of the backend and the executor should be the same and providing the executor at construction time makes that explicit.

xbrianh commented on dss/index/indexer.py:
```diff
@@ -35,9 +40,9 @@ def process_new_indexable_object(self, event: Mapping[str, Any]) -> None:
     def index_object(self, key):
```
There seems to be an (understandable) mixing of indexing logic and timing logic, but I wonder if these concerns could be separated.

For instance,
```
num_ops = indexer.load_data(*args)
check_time(num_ops * indexer.time_for_one_thing)
indexer.do_it()
```

mweiden replied:
To my understanding, @xbrianh is emphasizing where the logic lives. I think separating the concerns is possible by returning a data structure that defers the work once you know the size like the one below.

```python
class Task:

    def __init__(self, size, logic, *args, **kwargs):
        self.size = size
        self.logic = logic
        self.args = args
        self.kwargs = kwargs

    def submit(self):
        result = self.logic(*self.args, **self.kwargs)
        self.size = 0
        return result
```

So you could do something like.

```python
def foo(n):
    for i in range(n):
        time.sleep(1)

def some_insertion_logic():
    ...
    return Task(5, foo, 5)

# in a different (parent) module
print("Unexecuted task size is: " + str(task.size))

if task.size < 10:
    result = task.submit()
```

Just a thought.

hannes-ucsc replied:
To summarize our F2F conversation @mweiden, I don't think that the extra complexity of lifting  the unit of work (that's the design pattern your Task class follows) is warranted if we're moving the logic within the class and just one layer up in the call hierarchy.

The thing to keep in mind is that this PR already consolidates timing logic that was previously spread around more.

hannes-ucsc commented on dss/index/indexer.py:
```diff
@@ -48,25 +53,43 @@ def index_object(self, key):
+        logger.debug("Remaining indexing time is %fs", remaining_time)
+        return remaining_time
+
+    def _is_enough_time(self, num_operations):
```
Will do.

hannes-ucsc commented on dss/index/bundle.py:
```diff
@@ -64,14 +67,14 @@ def _read_file_infos(cls, replica: Replica, fqid: BundleFQID, manifest: JSON) ->
-                    if attempts < 5:
+                    # Requeue at the end, yielding to other files we may not have tried yet
+                    files_info.append((file_info, attempts + 1))
+                    if wait_time + len(files_info) * load_time < indexer.remaining_time:
```
Good point. Indexer depends on Bundle but Bundle should not depend on Indexer. In fact, if I add a `Indexer` type hint to this method, I'd have to import `Indexer` and that would result in a circular import.

However, this code *does* need to know the remaining run time on a continuing basis. One alternative to passing an `Indexer` instance is passing the lambda context directly. I thought that that was uglier because it exposes Bundle to something even further removed from its core concern.

The reason I had made peace with this ugliness is that it is going to go away once we either ensure that blobs are copied in order (#1130) or pursue @xbrianh's idea of retrying the entire indexer lambda using the @amork's DLQ mechanism.

hannes-ucsc commented:
So I ripped out the entire retry logic in `_load_file_infos()`. This eliminates the dependency of Bundle on Indexer, thereby making the coupling between the two unidirectional.

At the time @Bento007 and I implemented the retry logic in `_load_file_infos()` the reaper hadn't been merged yet. I think that the reaper is the perfect mechanism to pickup indexer lambda failures like those caused by the out-of-order syncing of the bundle manifest. Because that issue is so rare, it does not warrant the complexity we added, even if it avoids loading files twice or may lead to a successfully indexed bundle sooner.

That addresses one of your two big issues, @mweiden. The other issue of further separating the timing concern from main indexing concerns I find esoteric. As I laid out above, even *if* we were able to address this by introducing the unit of work pattern, the gain would be questionable. This PR is already a great improvement in that it consolidates the logic in a single class. I look forward to reviewing a PR that consolidates and isolates the timeout logic even further.
